src/server/hotspot_partition_calculator.cpp

```diff
@@ -34,67 +35,84 @@ DSN_DEFINE_int64("pegasus.collector",
     }
 }
 
 void hotspot_partition_calculator::data_analyse()
 {
-    dassert(_partition_stat_histories.back().size() == _hot_points.size(),
-            "partition counts error, please check");
-    std::vector<double> data_samples;
-    data_samples.reserve(_partition_stat_histories.size() * _hot_points.size());
-    auto temp_data = _partition_stat_histories;
-    double table_qps_sum = 0, standard_deviation = 0, table_qps_avg = 0;
-    int sample_count = 0;
-    while (!temp_data.empty()) {
-        for (const auto &partition_data : temp_data.front()) {
-            if (partition_data.total_qps - 1.00 > 0) {
-                data_samples.push_back(partition_data.total_qps);
-                table_qps_sum += partition_data.total_qps;
-                sample_count++;
+    dcheck_eq(_partitions_stat_histories.back().size(), _hot_points.size());
+    for (int data_type = 0; data_type <= 1; data_type++) {
+        // 0: READ_HOTSPOT_DATA; 1: WRITE_HOTSPOT_DATA
+        double table_qps_sum = 0, standard_deviation = 0, table_qps_avg = 0;
+        int sample_count = 0;
+        for (const auto &one_partition_stat_histories : _partitions_stat_histories) {
+            for (const auto &partition_stat : one_partition_stat_histories) {
+                if (partition_stat.total_qps[data_type] > 1.00) {
+                    table_qps_sum += partition_stat.total_qps[data_type];
+                    sample_count++;
+                }
             }
         }
-        temp_data.pop();
-    }
-    if (sample_count == 0) {
-        ddebug("_partition_stat_histories size == 0");
-        return;
-    }
-    table_qps_avg = table_qps_sum / sample_count;
-    for (const auto &data_sample : data_samples) {
-        standard_deviation += pow((data_sample - table_qps_avg), 2);
-    }
-    standard_deviation = sqrt(standard_deviation / sample_count);
-    const auto &anly_data = _partition_stat_histories.back();
-    for (int i = 0; i < _hot_points.size(); i++) {
-        double hot_point = (anly_data[i].total_qps - table_qps_avg) / standard_deviation;
-        // perf_counter->set can only be unsigned __int64
-        // use ceil to guarantee conversion results
-        hot_point = ceil(std::max(hot_point, double(0)));
-        _hot_points[i]->set(hot_point);
+
+        if (sample_count <= 1) {
+            ddebug("_partitions_stat_histories size <= 1");
+            return;
+        }
+        table_qps_avg = table_qps_sum / sample_count;
+        for (const auto &one_partition_stat_histories : _partitions_stat_histories) {
+            for (const auto &partition_stat : one_partition_stat_histories) {
+                if (partition_stat.total_qps[data_type] > 1.00) {
+                    standard_deviation +=
+                        pow((partition_stat.total_qps[data_type] - table_qps_avg), 2);
+                }
+            }
+        }
+        standard_deviation = sqrt(standard_deviation / (sample_count - 1));
```

Comment:
```suggestion
        standard_deviation = sqrt(standard_deviation / (sample_count - 1));
```

`(sample_count - 1)` Why sub 1?

Reply:
According https://en.wikipedia.org/wiki/Standard_deviation, the denominator in the sample standard deviation formula is N − 1.